core/frontend.ml

```diff
@@ -32,231 +40,267 @@ let show_post_frontend_ast
+       ; (module DesugarLAttributes)
+       ; (module LiftRecursive)
+       ; (module DesugarDatatypes)
+      |]
+
+  let run : Context.t -> ((module Untyped.S) -> Context.t -> 'a -> 'a Transform.Untyped.result) -> 'a -> 'a result
+    = fun context' select program ->
+    let module TU = Transform.Untyped in
+    let apply : 'a TU.result -> transformer -> 'a TU.result
+      = fun (TU.Result { program; state }) (module T) ->
+      select (module T) state program
+    in
+    let (TU.Result { state; program }) =
+      Array.fold_left apply TU.(return context' program) transformers
+    in
+    { context = state; program;
+      datatype = `Not_typed (* Slight abuse! *) }
+
+  let run_sentence : Context.t ->
+                     Sugartypes.sentence ->
+                     Sugartypes.sentence result
+    = fun context sentence ->
+    run context (fun (module T) -> T.Untyped.sentence) sentence
+
+  let run_program : Context.t ->
+                    Sugartypes.program ->
+                    Sugartypes.program result
+    = fun context program ->
+    run context (fun (module T) -> T.Untyped.program) program
+end
 
-  if Settings.get show_post_frontend_ast then
-    Debug.print ("Post-Frontend AST:\n" ^ Sugartypes.show_sentence result_sentence);
+(* The [Typeability_preserving] module runs transformation passes
+   which _must_ preserve typeability, i.e. the preimage and image of
+   the transform must be typeable. *)
+module Typeability_preserving = struct
+  type transformer = (module Typeable.S)
+
+  (* This functor constructs a conditional transformer from a given
+       transformer and a (side-effecting) condition. *)
+  module Conditional(T : sig
+               include Typeable.S
+               val condition : unit -> bool
+             end) = struct
+
+    module Typeable = struct
+      let program state program =
+        if T.condition ()
+        then T.Typeable.program state program
+        else Identity.Typeable.program state program
+
+      let sentence state sentence =
+        if T.condition ()
+        then T.Typeable.sentence state sentence
+        else Identity.Typeable.sentence state sentence
+    end
+  end
+
+  let only_if : bool Settings.setting -> (module Typeable.S) -> transformer
+    = fun setting (module T) ->
+    (module Conditional(struct include T let condition () = Settings.get setting end))
+
+  (* Collection of transformers. *)
+  let transformers : (string * transformer) array
+    = [| "cp", (module DesugarCP)
+       ; "inners", (module DesugarInners)
+       ; "session_execeptions", only_if
+                                  Basicsettings.Sessions.exceptions_enabled
+                                  (module DesugarSessionExceptions)
+       ; "processes", (module DesugarProcesses)
+       ; "fors", (module DesugarFors)
+       ; "regexes", (module DesugarRegexes)
+       ; "formlets", (module DesugarFormlets)
+       ; "pages", (module DesugarPages)
+       ; "funs", (module DesugarFuns) |]
```

Comment:
I wonder if it would be possible to have modules define their own name/description, rather than having them defined independently?

Reply:
Yes, that is a good idea, and should be easy to implement.